File: 정올/Intermediate_Coder/동적계획법/동전교환.py
n = int(input())
coin = list(map(int, input().split()))
w = int(input())

d = [-1] * (w + 1)
for c in coin:
    d[c] = 1

for i in range(1, w + 1):
    for c in coin:
        if c <= i and d[i - c] != -1:
            if d[i] != -1:
                d[i] = min(d[i], d[i - c] + 1)
            else:
                d[i] = d[i - c] + 1

if d[w] == -1:
    print("impossible")
else:
    print(d[w])

File: 정올/Intermediate_Coder/동적계획법/동전교환.cpp
#include <bits/stdc++.h>
using namespace std;

#define REP(i,s,e) for (int i = s; i < e; i++)
#define rep(i,e) for (int i = 0; i < e; i++)
#define INF 100000

int main() {
    ios_base::sync_with_stdio(false); cin.tie(NULL); cout.tie(NULL);

    int n, target;
    cin>>n;
    vector<int> coins(n);
    rep(i,n)
        cin>>coins[i];
    cin>>target;

    vector<int> dp(target+1);
    dp[0] = 0;
    REP(i,1,target+1){
        dp[i] = INF;
        for (auto c : coins)
            if (i-c >= 0)
                dp[i] = min(dp[i], dp[i-c]+1);
    }

    if (dp[target] >= INF) cout<<"impossible";
    else cout<<dp[target];

    return 0;
}
